src/instance.rs

```diff
@@ -724,6 +858,10 @@ impl<T> Py<T> {
         Py2(py, ManuallyDrop::new(self))
     }
 
+    pub(crate) fn attach_borrow<'a, 'py>(&'a self, py: Python<'py>) -> Py2Borrowed<'a, 'py, T> {
```

Comment:
Reposted from #3654: But doesn't attach_borrow sort of require the immutability guarantee of Py<PyBytes>::as_bytes now to avoid unsafely extending references into e.g. mutable pyclasses beyond the duration for which the GIL is held? Wouldn't this imply that attach_borrow needs to be unsafe?

Reply:
Can you elaborate a bit on a the safety violation you see? I might be missing something here.

I think you're right that `attach_borrow` isn't practically useful except in the `Py<PyBytes>` kind of situations, but I don't see how creating a `Py2Borrowed` with lifetime `'a` longer than `'py` is in itself problematic.

Reply:
I think this would mainly be relevant for Python implemented using interior mutability: Let's say we have `Py2<'py, T>` with `fn borrow<'a>(&'a self) -> RefWrapper<'a>` which set an internal borrow flag protected by the GIL. This is fine as `&'a Py2<'py, T>` requires that `'py` outlives `'a`, i.e. the GIL is held while the returned reference wrapper is alive.

But now, I can use `attach_borrow` to add `fn borrow<'a>(&'a self) -> RefWrapper<'a>` to `Py<T>` giving me a reference wrapper which will reset the internal borrow flag after the GIL was already released, racing with other threads having the GIL and trying to get a borrow themselves.

Of course, this is not public so we could just declare "don't do that", but I think we should try to help ourselves avoid mistakes as well. Having a reference into an object on the Python heap that outlives holding the GIL is so unexpected/fragile, that I would prefer making the constructor of such a type unsafe.

Reply:
If I follow your wording correctly, then the construct which concerns you is this:

```rust
impl Py<T> {
    fn borrow<'a>(&'a self, py: Python<'_>) -> RefWrapper<'a> {
         self.attach_borrow(py).borrow()
    }
}
```

Based on the lessons I've learned wrangling with these lifetimes trying to get this all right, I'm reasonably sure we're ok and that this won't compile.

The reason is that `self.attach_borrow(py).borrow()` would create a temporary `Py2Borrowed`, which we then need to deref to access the `borrow` method of `Py2`. Therefore the lifetime of the returned `RefWrapper` is not long enough, as it would be limited to the deref of the temporary.

To make the above `fn borrow` compile, I think we'd need the following method on `Py2Borrowed` to consume the temporary and spit out an appropriately long `RefWrapper`:

```rust
impl<'a, T> Py2Borrowed<'a, '_, T> {
    fn borrow(self) -> RefWrapper<'a> {
        /* ... */
    }
}
```

And I think we should be able to catch in review that that method can't be sound, because it spits out a `RefWrapper` ignoring the GIL lifetime.

Reply:
Yes, you are right, I was missing the `as_bytes` method directly on `Py2Borrow` which is required to make this possible. And indeed without it, `error[E0515]: cannot return value referencing temporary value` prevents anything bad from happening.